omnibus/cookbooks/omnibus-supermarket/spec/libraries/config_spec.rb

```diff
@@ -75,10 +92,146 @@
+      end
+
+    end
+  end
+
+  describe '#maybe_turn_on_fips' do
+    let(:node) { Chef::Node.new() }
+
+    context 'with the default setting' do
+      before(:each) do
+        node.default['supermarket']['fips_enabled'] = nil
+      end
+
+      context 'and a kernel without FIPS enabled (the usual path)' do
+        before(:each) do
+          allow(described_class).to receive(:fips_enabled_in_kernel?).and_return(false)
+        end
+
+        it 'does not enable FIPS' do
+          described_class.maybe_turn_on_fips(node)
+          expect(node['supermarket']['fips_enabled']).to be false
+        end
+      end
+
+      context 'and a kernel with FIPS enabled' do
+        before(:each) do
+          allow(described_class).to receive(:fips_enabled_in_kernel?).and_return(true)
+        end
+
+        it 'enables FIPS and logs messages about it' do
+          expect(Chef::Log).to receive(:warn).with('Detected FIPS-enabled kernel; enabling FIPS 140-2 for Supermarket services.')
+          described_class.maybe_turn_on_fips(node)
+          expect(node['supermarket']['fips_enabled']).to be true
+        end
+      end
+    end
+
+    context 'with fips_enabled set to true' do
+      before(:each) do
+        node.consume_attributes('supermarket' => { 'fips_enabled' => true })
+      end
+
+      it 'enables FIPS and logs messages about it' do
+        expect(Chef::Log).to receive(:warn).with('Overriding FIPS detection: FIPS 140-2 mode is ON.')
+        described_class.maybe_turn_on_fips(node)
+        expect(node['supermarket']['fips_enabled']).to be true
+      end
+    end
+
+    context 'with fips_enabled set to false' do
+      before(:each) do
+        node.consume_attributes('supermarket' => { 'fips_enabled' => false })
+      end
+
+      context 'and a kernel without FIPS enabled' do
+        before(:each) do
+          allow(described_class).to receive(:fips_enabled_in_kernel?).and_return(false)
+        end
+
+        it 'does not enable FIPS' do
+          described_class.maybe_turn_on_fips(node)
+          expect(node['supermarket']['fips_enabled']).to be false
+        end
+      end
+
+      context 'and a kernel with FIPS enabled' do
+        before(:each) do
+          allow(described_class).to receive(:fips_enabled_in_kernel?).and_return(true)
+        end
+
+        it 'enables FIPS anyway and logs messages about why you gotta FIPS when it is enabled in the kernel' do
+          expect(Chef::Log).to receive(:warn).with('Detected FIPS-enabled kernel; enabling FIPS 140-2 for Supermarket services.')
+          expect(Chef::Log).to receive(:warn).with('fips_enabled was set to false; ignoring this and setting to true or else Supermarket services will fail with crypto errors.')
+          described_class.maybe_turn_on_fips(node)
+          expect(node['supermarket']['fips_enabled']).to be true
+        end
+      end
+    end
+
+    context 'with fips_enabled set to anything other than nil or boolean true/false' do
```

Comment:
I didn't comment in the check above, but I am wondering why were' supporting non-boolean values with a default behavior? It seems opaque to the user.

Reply:
This came out of some exploring admin expectations and some prior art with FIPSing in core Chef.

@cattywampus and I were confused at first when testing this in a Test Kitchen VM. We had set fips_enabled to `true` through TK's attributes and still weren't seeing the Supermarket services turn FIPS on. Turns out it was the explicit check for boolean `node['supermarket']['fips_enabled'] == true` in the service templates and TK was setting it to the String `"true"`.

Looking at how FIPS is detected, disabled, or enabled shows that we have previously used "literally anything else" as a way to enable FIPS, so we did that. Downstream code in the cookbook can rely on `fips_enabled` being a Proper Boolean while we can accommodate admins giving a string form of truth to the config.

Reply:
Ugh. Okay thanks.